bindings/js/api.js

```diff
@@ -0,0 +1,283 @@
+  in that case we check dynamically whether they have the correct length), or
+  have a variable length (and we have to pass that length as an additional
+  parameter to WebAssembly).
+
+  In order to match the Javascript API with the actual calls to WebAssembly functions,
+  we have to describe the correspondance between the two in the `api.json` file.
+
+  The scheme of the JSON file is the following :
+  - `module`, this module name will be shown in the JS API
+    - `function`, this function name will be shown in the JS API
+      - `module`, the name of the WebAssembly file where to find the function
+      - 'name', the name of the WebAssembly export corresponding to the function
+      - 'args', the list of the WebAssembly arguments expected by the function
+        - 'name', the name of the argument which will be shown in the JS Doc
+        - 'kind', either `input` or `output` of the function
+        - 'type', either 'int' or 'boolean' or 'buffer'
+        - 'size', either an integer of a string which is the 'name' of another
+          argument of type 'int'
+        - 'interface_index', for all `input` that should appear in JS, position
+          inside the argument list of the JS function
+        - 'tests', a list of values for this arguments, each value corresponding
+          to a different test case
+      - 'return', the return type of the WebAssembly function
+  */
+
+  const CheckIfByteArray = function(candidate, length, name) {
+    if (!(typeof(candidate) === "object") ||
+      !(candidate.length === length) ||
+      !(candidate.constructor === Uint8Array)
+    ) {
+      throw new Error(
+        "name: Please ensure the argument " + candidate + " is a " + length + "-bytes Uint8Array."
+      );
+    }
+  };
+
+  const copy_array_to_stack = function(array) {
+    let pointer = loader.reserve(Module.Kremlin.mem, array.length);
+    (new Uint8Array(Module.Kremlin.mem.buffer)).set(array, pointer);
+    return pointer;
+  };
+
+  const read_memory = function(ptr, len) {
+    var result = new ArrayBuffer(len);
+    (new Uint8Array(result).set(new Uint8Array(Module.Kremlin.mem.buffer)
+      .subarray(ptr, ptr + len)));
+    return new Uint8Array(result);
+  };
+
+  const callWithProto = function(proto, args, loc_name) {
+    var expected_args_number = proto.args.filter(arg => arg.interface_index !== undefined).length;
+    if (args.length != expected_args_number) {
+      throw Error("wrong number of arguments to call the F*-wasm function " + loc_name + ": expected " + expected_args_number + ", got " + args.length)
+    }
+    let memory = new Uint32Array(Module.Kremlin.mem.buffer);
+    let sp = memory[0];
+    var var_lengths = {};
+    // Populating the variable length arguments by retrieving buffer lengths
+    proto.args.map((arg) => {
+      if (arg.type === "buffer") {
+        if ((typeof arg.size === "string") && (arg.interface_index !== undefined)) {
```

Comment:
not sure there's much to do about it, but wondering whether the checks for well-formedness can be shared between the validator and this?

Reply:
So the lines of code that you're highlighting here are not checks _per se_. Here we have to case on the type of the arguments because buffers are allocated on the wasm memory whereas base types can be passed directly. Generally all the code in `api.js` assumes that the data it is fed has passed the well-formedness check.